src/QsCompiler/CompilationManager/EditorCompletion.cs

```diff
@@ -0,0 +1,686 @@
+        private static string ResolveNamespaceAlias(
+            FileContentManager file, CompilationUnit compilation, Position position, string alias)
+        {
+            if (file == null)
+                throw new ArgumentNullException(nameof(file));
+            if (compilation == null)
+                throw new ArgumentNullException(nameof(compilation));
+            if (position == null)
+                throw new ArgumentNullException(nameof(position));
+            if (alias == null)
+                throw new ArgumentNullException(nameof(alias));
+
+            var nsName = file.TryGetNamespaceAt(position);
+            if (nsName == null)
+                return alias;
+            return compilation.GlobalSymbols.TryResolveNamespaceAlias(
+                NonNullable<string>.New(alias), NonNullable<string>.New(nsName), file.FileName) ?? alias;
+        }
+
+        /// <summary>
+        /// Gets the token index and corresponding code fragment at, or the closest before, the given position in the
+        /// file. Returns the relationship between the found fragment and the fragment that would exist directly at the
+        /// given position, which may be different from <see cref="FragmentRelationship.Self"/> if the given position is
+        /// not part of any fragment.
+        /// <para/>
+        /// If <see cref="FragmentRelationship.Missing"/> is returned, then both the token index and code fragment out
+        /// parameters are set to null.
+        /// </summary>
+        /// <exception cref="ArgumentNullException">Thrown when any argument is null.</exception>
+        private static FragmentRelationship GetFragmentAtOrBefore(
+            FileContentManager file,
+            Position position,
+            out CodeFragment.TokenIndex index,
+            out CodeFragment fragment)
+        {
+            if (file == null)
+                throw new ArgumentNullException(nameof(file));
+            if (position == null)
+                throw new ArgumentNullException(nameof(position));
+
+            var lineNumber = position.Line;
+            var tokens = file.GetTokenizedLine(lineNumber);
+
+            // If the current line is empty, find the last non-empty line before it.
+            while (tokens.IsEmpty && lineNumber > 0)
+                tokens = file.GetTokenizedLine(--lineNumber);
+
+            var indexNumber = tokens.TakeWhile(ContextBuilder.TokensUpTo(position)).Count() - 1;
+            if (indexNumber == -1)
+            {
+                index = null;
+                fragment = null;
+                return FragmentRelationship.Missing;
+            }
+
+            index = new CodeFragment.TokenIndex(file, lineNumber, indexNumber);
+            fragment = index.GetFragment();
+            if (fragment.FollowedBy != CodeFragment.MissingDelimiter &&
+                GetDelimitingCharPosition(file, fragment).IsSmallerThan(position))
+            {
+                switch (fragment.FollowedBy)
```

Comment:
It may be better to check the indentation level that is assigned to each fragment, to avoid having to hardcode the meaning of '{','}',';' here.